File: trunk/source/Result.cpp
#include "stdafx.h"
/*
* Copyright (c) 2007-2012 SlimDX Group
* 
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
* 
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
* 
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

#include <windows.h>
#include <dxerr.h>

#include "Configuration.h"
#include "Result.h"
#include "SlimDXException.h"

using namespace System;
using namespace System::Globalization;
using namespace System::Collections;

namespace SlimDX
{
	Result::Result( int hr )
	: m_Code( hr )
	{
	}
	
	int Result::Code::get()
	{
		return m_Code;
	}
	
	String^ Result::Name::get()
	{
		if( m_Info == nullptr )
		{
			m_Info = gcnew ResultInfo();
		}

		if( m_Info->Name == nullptr )
		{
			m_Info->Name = gcnew String( DXGetErrorString( m_Code ) );
		}

		return m_Info->Name;
	}

	String^ Result::Description::get()
	{
		if( m_Info == nullptr )
		{
			m_Info = gcnew ResultInfo();
		}

		if( m_Info->Description == nullptr )
		{
			m_Info->Description = gcnew String( DXGetErrorDescription( m_Code ) );
		}
		
		return m_Info->Description;
	}
	
	SortedList^ Result::Data::get()
	{
		if( m_Info == nullptr )
		{
			m_Info = gcnew ResultInfo();
		}

		return m_Info->Data;
	}

	void Result::Data::set( SortedList^ value )
	{
		if( m_Info == nullptr )
		{
			m_Info = gcnew ResultInfo();
		}

		m_Info->Data = value;
	}

	bool Result::IsSuccess::get()
	{
		return SUCCEEDED( m_Code );
	}
	
	bool Result::IsFailure::get()
	{
		return FAILED( m_Code );
	}
	
	void Result::BreakIfDebugging()
	{
		System::Diagnostics::Debugger::Break();
	}

	generic< typename T >
	void Result::Throw( Object^ dataKey, Object^ dataValue )
	{
		SlimDXException^ ex = safe_cast<SlimDXException^>( Activator::CreateInstance( T::typeid, m_Last ) );
		if( dataKey != nullptr )
			ex->Data->Add(dataKey, dataValue);

		throw ex;
	}

	generic< typename T >
	Result Result::Record( int hr, bool failed, Object^ dataKey, Object^ dataValue )
	{
		m_LastCode = hr;
		if(!failed)
			return Result( hr );

		m_Last = Result( hr );
		if( dataKey != nullptr )
		{
			m_Last.Data = gcnew SortedList();
			m_Last.Data->Add( dataKey, dataValue );
		}

		ResultWatchFlags flags;
		if( Configuration::TryGetResultWatch( m_Last, flags ) )
		{
			if( flags == ResultWatchFlags::AlwaysIgnore )
				return m_Last;

			if( static_cast<int>( flags & ResultWatchFlags::Break ) != 0 )
				BreakIfDebugging();

			if( static_cast<int>( flags & ResultWatchFlags::Throw ) != 0 )
				Throw<T>(dataKey, dataValue);
		}

		if( Configuration::ThrowOnError )
			Throw<T>(dataKey, dataValue);

		return m_Last;
	}

	generic< typename T >
	Result Result::Fail( int hr, Object^ dataKey, Object^ dataValue )
	{
		return Record<T>( hr, true, dataKey, dataValue );
	}

	generic< typename T >
	Result Result::Record( int hr, Object^ dataKey, Object^ dataValue )
	{
		return Record<T>( hr, hr < 0, dataKey, dataValue );
	}
	
	Result Result::Last::get()
	{
		if( SUCCEEDED(m_LastCode) )
			return Result( m_LastCode );
		else
			return m_Last;
	}
	
	bool Result::operator == ( Result left, Result right )
	{
		return Result::Equals( left, right );
	}
	
	bool Result::operator != ( Result left, Result right )
	{
		return !Result::Equals( left, right );
	}

	String^ Result::ToString()
	{
		return String::Format( CultureInfo::CurrentCulture, "{0}: {1} ({2})", Name, Description, Code );
	}

	int Result::GetHashCode()
	{
		return m_Code.GetHashCode();
	}

	bool Result::Equals( Object^ value )
	{
		if( value == nullptr )
			return false;

		if( value->GetType() != GetType() )
			return false;

		return Equals( safe_cast<Result>( value ) );
	}

	bool Result::Equals( Result value )
	{
		return ( Code == value.Code );
	}

	bool Result::Equals( Result% value1, Result% value2 )
	{
		return ( value1.Code == value2.Code );
	}
}

File: branches/v2/Source/SlimDX/Result.cs
// Copyright (c) 2007-2010 SlimDX Group
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
using System;
using System.Reflection;
using System.Runtime.InteropServices;

namespace SlimDX
{
    /// <summary>
    /// Result structure for COM methods.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct Result : IEquatable<Result>
    {
        private int _code;
        private static MethodInfo _methodGetErrorDescription;

        public Result(int code)
        {
            _code = code;
        }

        public Result(uint code)
        {
            _code = unchecked((int)code);
        }

        public int Code
        {
            get { return _code; }
        }

        public bool Success
        {
            get { return Code >= 0; }
        }

        public bool Failure
        {
            get { return Code < 0; }
        }

        public static implicit operator Result(int result)
        {
            return new Result(result);
        }

        public bool Equals(Result other)
        {
            return this.Code == other.Code;
        }

        public override string ToString()
        {
            if (_methodGetErrorDescription != null)
                return _methodGetErrorDescription.Invoke(null, new object[] {_code}).ToString();
            return string.Format("Unknown error (HRESULT = 0x{0:X})", _code);
        }

        public void CheckError()
        {
            if (_code < 0)
            {
                throw new SlimDX2Exception(this);
            }
        }

        static Result()
        {
            _methodGetErrorDescription = null;
            try
            {
                string prefix = typeof (Result).Namespace;
                Assembly assembly = Assembly.LoadFrom(    prefix + ".Error.dll");
                Type type = assembly.GetType(prefix + ".ErrorManager");
                _methodGetErrorDescription = type.GetMethod("GetErrorMessage", BindingFlags.Static | BindingFlags.Public);
            }
            catch (Exception ex)
            {                
            }
        }

        public static Result Ok = new Result(unchecked((int)0x00000000));
        public static Result Abord = new Result(unchecked((int)0x80004004));
        public static Result AccessDenied = new Result(unchecked((int)0x80070005));
        public static Result Fail = new Result(unchecked((int)0x80004005));
        public static Result Handle = new Result(unchecked((int)0x80070006));
        public static Result InvalidArg = new Result(unchecked((int)0x80070057));
        public static Result NoInterface = new Result(unchecked((int)0x80004002));
        public static Result NotImplemented = new Result(unchecked((int)0x80004001));
        public static Result OutOfMemory = new Result(unchecked((int)0x8007000E));
        public static Result InvalidPointer = new Result(unchecked((int)0x80004003));
        public static Result UnexpectedFailure = new Result(unchecked((int)0x8000FFFF));
    }
}